lms/djangoapps/program_enrollments/api/v1/views.py

```diff
@@ -124,3 +135,196 @@ def get(self, request, program_key=None):
+            student_key = enrollment["student_key"]
+            if student_key in results and results[student_key] == CourseEnrollmentResponseStatuses.DUPLICATED:
+                continue
+            results[student_key] = self.enroll_learner_in_course(enrollment, program_enrollments)
+
+        good_count = sum([1 for _, v in results.items() if v not in CourseEnrollmentResponseStatuses.ERROR_STATUSES])
+        if not good_count:
+            return Response(results, status.HTTP_422_UNPROCESSABLE_ENTITY)
+        if good_count != len(results):
+            return Response(results, status.HTTP_207_MULTI_STATUS)
+        else:
+            return Response(results)
+
+    def check_enrollment_request(self, enrollment, seen_student_keys):
+        """
+        Checks that the given enrollment record is valid and hasn't been duplicated
+        """
+        student_key = enrollment['student_key']
+        if student_key in seen_student_keys:
+            return CourseEnrollmentResponseStatuses.DUPLICATED
+        seen_student_keys.add(student_key)
+        enrollment_serializer = ProgramCourseEnrollmentRequestSerializer(data=enrollment)
+        try:
+            enrollment_serializer.is_valid(raise_exception=True)
+        except ValidationError as e:
+            if enrollment_serializer.has_invalid_status():
+                return CourseEnrollmentResponseStatuses.INVALID_STATUS
+            else:
+                raise e
+
+    def get_existing_program_enrollments(self, program_uuid, enrollments):
+        """
+        Parameters:
+            - enrollments: A list of enrollment requests
+        Returns:
+            - Dictionary mapping all student keys in the enrollment requests
+              to that user's existing program enrollment <self.program>
+        """
+        external_user_keys = [e["student_key"] for e in enrollments]
+        existing_enrollments = ProgramEnrollment.objects.filter(
+            external_user_key__in=external_user_keys,
+            program_uuid=program_uuid,
+        )
+        existing_enrollments = existing_enrollments.prefetch_related('program_course_enrollments')
+        return {enrollment.external_user_key: enrollment for enrollment in existing_enrollments}
+
+    def enroll_learner_in_course(self, enrollment_request, program_enrollments):
+        """
+        Attempts to enroll the specified user into the course as a part of the
+         given program enrollment with the given status
+
+        Returns the actual status
+        """
+        student_key = enrollment_request['student_key']
+        try:
+            program_enrollment = program_enrollments[student_key]
+        except KeyError:
+            return CourseEnrollmentResponseStatuses.NOT_IN_PROGRAM
+        if program_enrollment.get_program_course_enrollment(self.course_key):
+            return CourseEnrollmentResponseStatuses.CONFLICT
+
+        enrollment_status = enrollment_request['status']
+        course_enrollment = None
```

Comment:
Everything after this line is data logic, so I think it might be more suited to living `models.py`, possible as a classmethod on ProgramCourseEnrollment (i.e., `def ProgramCourseEnrollment.enroll(cls, ...`)

Reply:
cool, i actually never considered putting it there but now that you mention it it does make sense